tools/xdp/s2n-quic-xdp/src/io/tx.rs

```diff
@@ -0,0 +1,193 @@
+            occupied,
+            free,
+            umem,
+            encoder,
+            is_full: false,
+        }
+    }
+
+    /// Consumes the TX endpoint into the inner channels
+    ///
+    /// This is used for internal tests only.
+    #[cfg(test)]
+    pub fn consume(self) -> (Free, Occupied) {
+        (self.free, self.occupied)
+    }
+}
+
+impl tx::Tx for Tx {
+    type PathHandle = path::Tuple;
+    type Queue = Queue<'static>;
+    type Error = spsc::ClosedError;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context) -> Poll<Result<(), Self::Error>> {
+        // poll both channels to make sure we can make progress in both
+        let free = self.free.poll_slice(cx);
+        let occupied = self.occupied.poll_slice(cx);
+
+        ready!(free)?;
+        ready!(occupied)?;
+
+        // we only need to wake up if the queue was previously completely filled up
+        if !self.is_full {
+            return Poll::Pending;
+        }
+
+        Poll::Ready(Ok(()))
+    }
+
+    #[inline]
+    fn queue<F: FnOnce(&mut Self::Queue)>(&mut self, f: F) {
+        let this: &'static mut Self = unsafe {
+            // Safety: As noted in the [transmute examples](https://doc.rust-lang.org/std/mem/fn.transmute.html#examples)
+            // it can be used to temporarily extend the lifetime of a reference. In this case, we
+            // don't want to use GATs until the MSRV is >=1.65.0, which means `Self::Queue` is not
+            // allowed to take generic lifetimes.
+            //
+            // We are left with using a `'static` lifetime here and encapsulating it in a private
+            // field. The `Self::Queue` struct is then borrowed for the lifetime of the `F`
+            // function. This will prevent the value from escaping beyond the lifetime of `&mut
+            // self`.
+            //
+            // See https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9a32abe85c666f36fb2ec86496cc41b4
+            //
+            // Once https://github.com/aws/s2n-quic/issues/1742 is resolved this code can go away
+            core::mem::transmute(self)
+        };
+
+        let mut free = this.free.slice();
+        let mut occupied = this.occupied.slice();
+
+        // if we were full, then try to synchronize the peer's queues
+        if this.is_full {
+            let _ = free.sync();
+            let _ = occupied.sync();
+        }
+
+        // update our full status
+        this.is_full = free.is_empty() || occupied.capacity() == 0;
+
+        let umem = &mut this.umem;
+        let encoder = &mut this.encoder;
+        let is_full = &mut this.is_full;
+
+        let mut queue = Queue {
+            free,
+            occupied,
+            umem,
+            encoder,
+            is_full,
+        };
+        f(&mut queue);
+    }
+
+    #[inline]
+    fn handle_error<E: event::EndpointPublisher>(self, _error: Self::Error, _events: &mut E) {
+        // no-op, just shut down
```

Comment:
why is it a no-op?

Reply:
I can expand in the comment but the only reason we'd shut down is if one of the channels closed on the other side. That would be due to a panic or the endpoint being dropped.